src/sandstorm/appid-replacements.c++

```diff
@@ -0,0 +1,96 @@
+// Sandstorm - Personal Cloud Sandbox
+// Copyright (c) 2016 Sandstorm Development Group, Inc. and contributors
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "appid-replacements.h"
+#include "id-to-text.h"
+
+namespace sandstorm {
+
+void applyAppidReplacements(
+    kj::ArrayPtr<kj::byte> appId, kj::ArrayPtr<const kj::byte> packageId,
+    capnp::List<spk::AppIdReplacement>::Reader replacements) {
+  // Given an input app ID that was just verified to have signed the given package ID,
+  // check appid-replacements.capnp to see if the app ID is revoked (throws exception) or the
+  // package should be treated as some other app (replaces appId).
+
+  // The logic here is slightly weird because the replacement list is organized into events --
+  // which makes it easier for people modifying it to understand what to do -- rather than into
+  // rules -- which could be more directly processed. Each event introduces one or two rules:
+  // a revocation of an original key (except for grandfathered whitelist) and a mapping of a
+  // replacement key to and original key.
+
+  KJ_REQUIRE(appId.size() == APP_ID_BYTE_SIZE);
+  KJ_REQUIRE(packageId.size() == PACKAGE_ID_BYTE_SIZE);
+
+  auto appidStr = appIdString(appId);
+  auto pkgidStr = packageIdString(packageId);
+
+  // First check if this app ID is revoked.
+  for (auto item: replacements) {
+    if (item.getOriginal() == appidStr) {
+      // The app ID matches the `original` of this entry. Check if it was revoked.
+      if (item.hasRevokeExceptPackageIds()) {
+        // This app ID is revoked, except for specific package IDs...
+        for (auto allowed: item.getRevokeExceptPackageIds()) {
+          if (pkgidStr == allowed) {
+            goto checkReplacements;  // Need to break outer loop...
```

Comment:
This is a totally reasonable usage of `goto`, but man, I haven't seen one of these in ages! :D